janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/addedrelations/SimpleAddedRelations.java

```diff
@@ -14,46 +14,74 @@
 
 package org.janusgraph.graphdb.transaction.addedrelations;
 
-import com.google.common.base.Predicate;
 import org.janusgraph.graphdb.internal.InternalRelation;
 
-import java.util.ArrayList;
+import com.carrotsearch.hppc.ObjectHashSet;
+import com.carrotsearch.hppc.ObjectSet;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import javax.annotation.Nonnull;
+
+import java.util.AbstractCollection;
 import java.util.Collection;
-import java.util.List;
+import java.util.Collections;
+import java.util.Iterator;
 
 /**
  * @author Matthias Broecheler (me@matthiasb.com)
  */
 
-public class SimpleAddedRelations extends ArrayList<InternalRelation> implements AddedRelationsContainer {
-
+public class SimpleAddedRelations implements AddedRelationsContainer {
     private static final int INITIAL_ADDED_SIZE = 10;
+    private final ObjectSet<InternalRelation> container;
 
     public SimpleAddedRelations() {
-        super(INITIAL_ADDED_SIZE);
+        this.container = new ObjectHashSet<>(INITIAL_ADDED_SIZE);
     }
 
     @Override
     public boolean add(InternalRelation relation) {
-        return super.add(relation);
+        container.add(relation);
+        return true;
     }
 
     @Override
     public boolean remove(InternalRelation relation) {
-        return super.remove(relation);
+        container.removeAll(relation);
+        return true;
```

Comment:
Why `removeAll` instead of `remove`?
Why it returns `true`? Should it be `return remove(relation)`?

Reply:
`removeAll` is alias for `remove` for set based, so I fix it

Reply:
Thanks for explanation. I see it. Not necessary but you can extend `ObjectHashSet` class and overwrite necessary methods instead of making a proxy for `ObjectSet` (something like here https://github.com/JanusGraph/janusgraph/pull/1701#discussion_r323361562). Then you will have `remove` method and you won't need to overwrite `isEmpty` method.
For example:
```
@Override
public boolean remove(InternalRelation relation) {
  super.remove(relation);
  return true;
}
```

This is up to you. You may leave it as a proxy or make it as an extended class.

Reply:
I now think that proxy might be a better option. Right now we can safely extend `ObjectHashSet` because it uses `this.remove(..)` inside `removeAll(..)` method but if it changes the implementation for some reason to `remove(..)` instead of `this.remove(..)` we might have problems with such implementation because we overwrite `remove(..)` method of `ObjectHashSet`.
So, that said, I think your current implementation is good because it decreases potential future bugs. I now think that we should not extend `ObjectHashSet` as I suggested earlier.